src/lib.rs

```diff
@@ -409,32 +399,53 @@ const DERIVE_WHERE: &str = "derive_where";
 ///
 /// [`Zeroize`]: https://docs.rs/zeroize/latest/zeroize/trait.Zeroize.html
 /// [`ZeroizeOnDrop`]: https://docs.rs/zeroize/latest/zeroize/trait.ZeroizeOnDrop.html
-#[proc_macro_derive(DeriveWhere, attributes(derive_where))]
+#[proc_macro_attribute]
 #[cfg_attr(feature = "nightly", allow_internal_unstable(core_intrinsics))]
-pub fn derive_where(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
-	match derive_where_internal(input.into()) {
-		Ok(output) => output.into(),
-		Err(error) => error.into_compile_error().into(),
-	}
-}
-
-/// Internal derive function for handling errors.
-fn derive_where_internal(input: TokenStream) -> Result<TokenStream> {
+pub fn derive_where(
+	attr: proc_macro::TokenStream,
+	original_input: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+	let attr: TokenStream = attr.into();
+	let input: TokenStream = original_input.clone().into();
+	let input = quote! {
+		#[derive_where(#attr)]
+		#input
+	};
 	// Save `Span` before we consume `input` when parsing it.
 	let span = input.span();
-	let item = syn::parse2::<DeriveInput>(input).expect("derive on unparsable item");
-
-	let Input {
-		derive_wheres,
-		generics,
-		item,
-	} = Input::from_input(span, &item)?;
+	let item = match syn::parse2::<DeriveInput>(input) {
+		Ok(item) => item,
+		Err(error) => {
+			return iter::once(original_input)
+				.chain(iter::once(proc_macro::TokenStream::from(
+					error.into_compile_error(),
+				)))
+				.collect();
+		}
+	};
+	let cleaned_item = input_without_derive_where_attributes(item.clone());
 
-	Ok(derive_wheres
-		.iter()
-		.flat_map(|derive_where| iter::repeat(derive_where).zip(&derive_where.traits))
-		.map(|(derive_where, trait_)| generate_impl(derive_where, trait_, &item, generics))
-		.collect())
+	match { Input::from_input(span, &item) } {
```

Comment:
I think the issue is actually here. This span references something that is probably dead, because we modified it? You can try taking the span from the cleaned item and see what happens.

Reply:
seams to have helped changing this!